cmd/clusterctl/client/tree/discovery.go

```diff
@@ -93,37 +107,124 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 
 		if visible {
 			if machineInfra, err := external.Get(ctx, c, &m.Spec.InfrastructureRef, cluster.Namespace); err == nil {
-				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(true))
+				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(!options.Echo))
 			}
 
 			if machineBootstrap, err := external.Get(ctx, c, m.Spec.Bootstrap.ConfigRef, cluster.Namespace); err == nil {
-				tree.Add(m, machineBootstrap, ObjectMetaName("BootstrapConfig"), NoEcho(true))
+				tree.Add(m, machineBootstrap, ObjectMetaName("BootstrapConfig"), NoEcho(!options.Echo))
 			}
 		}
 	}
 
 	controlPlaneMachines := selectControlPlaneMachines(machinesList)
 	for i := range controlPlaneMachines {
 		cp := controlPlaneMachines[i]
-		addMachineFunc(controlPLane, cp)
+		addMachineFunc(controlPlane, cp)
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(machinesList.Items) == len(controlPlaneMachines) && len(machinePoolList.Items) == 0 {
 		return tree, nil
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
 	tree.Add(cluster, workers)
 
+	err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	addMachinePoolsToObjectTree(ctx, c, cluster, workers, machinePoolList, tree, options)
+
+	// Handles orphan machines.
+	if len(machineMap) < len(machinesList.Items) {
+		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+		tree.Add(workers, other)
+
+		for i := range machinesList.Items {
+			m := &machinesList.Items[i]
+			if _, ok := machineMap[m.Name]; ok {
+				continue
+			}
+			addMachineFunc(other, m)
+		}
+	}
+
+	return tree, nil
+}
+
+func constructPlaceholderObject(objectRef *corev1.ObjectReference) *unstructured.Unstructured {
+	placeholder := unstructured.Unstructured{}
+	placeholder.SetKind(objectRef.Kind)
+	placeholder.SetNamespace(objectRef.Namespace)
+	placeholder.SetName(objectRef.Name)
+	placeholder.SetAPIVersion(objectRef.APIVersion)
+	placeholder.SetUID(types.UID(objectRef.Kind + "/" + objectRef.Namespace + "/" + objectRef.Name))
```

Comment:
I had to set a UID in the placeholder objects since it's used as the object key in `tree.go`. I don't believe we can get the actual UID on the resource without calling a `Get()` call so I filled it in with some temporary values. Do you think these values are fine?